### PR TITLE
Refactored er_class

### PR DESCRIPTION
**Pull request by Vitaly-Protasov**
Solved #510

### CODE REVIEW

File: aibolit/patterns/er_class/er_class.py [hunk 1]
```diff
@@ -1,14 +1,15 @@
-import javalang
-
+from aibolit.ast_framework import ASTNodeType, AST
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:1]: Add header with license. I am using 'File Header Comment' extension to simplify this task.

File: aibolit/patterns/er_class/er_class.py [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
 from aibolit.utils.ast_builder import build_ast
+from typing import List
 
 
 class ErClass:
 
     def __init__(self):
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:8]: I think we can remove empty constructors.

File: aibolit/patterns/er_class/er_class.py [hunk 3]
```diff
@@ -23,5 +24,14 @@ def value(self, filename: str):
                    'producer',
                    'holder',
                    'interceptor')
-        tree = build_ast(filename).filter(javalang.tree.ClassDeclaration)
-        return [node._position.line for _, node in tree if [n for n in classes if n in node.name.lower()] != []]
+        ast = AST.build_from_javalang(build_ast(filename))
+        class_decls = list(ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION))
+        for node in class_decls:
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:29]:
```suggestion
        for node in ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION):
```
And remove the line above. This way we would not create list of all nodes, but iterate one by one, without storing them all in memory.

File: aibolit/patterns/er_class/er_class.py [hunk 4]
```diff
@@ -1,14 +1,15 @@
-import javalang
-
+from aibolit.ast_framework import ASTNodeType, AST
 from aibolit.utils.ast_builder import build_ast
+from typing import List
 
 
 class ErClass:
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:6]: Ununderstandable name. Better be changed. Also a comment describing a pattern in details may be added to this class.

File: aibolit/patterns/er_class/er_class.py [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 
     def __init__(self):
         pass
 
-    def value(self, filename: str):
+    def value(self, filename: str) -> List[int]:
+        lines: List[int] = []
         classes = ('manager',
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:13]: Might be moved to class static field.

Comment (aravij) [reply aibolit/patterns/er_class/er_class.py:13]: Also might be renamed to something like `forbiden_words_in_class_names`.

File: aibolit/patterns/er_class/er_class.py [hunk 6]
```diff
@@ -23,5 +24,14 @@ def value(self, filename: str):
...
                    'holder',
                    'interceptor')
-        tree = build_ast(filename).filter(javalang.tree.ClassDeclaration)
-        return [node._position.line for _, node in tree if [n for n in classes if n in node.name.lower()] != []]
+        ast = AST.build_from_javalang(build_ast(filename))
+        class_decls = list(ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION))
+        for node in class_decls:
+            names = []
+            class_name = node.name.lower()
+            for name in classes:
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:32]: Maybe something like this:
```python
if any(forbiden_word in class_name for forbiden_words in forbiden_words_in_class_names):
```

File: aibolit/patterns/er_class/er_class.py [hunk 7]
```diff
@@ -23,5 +24,14 @@ def value(self, filename: str):
                    'producer',
                    'holder',
                    'interceptor')
-        tree = build_ast(filename).filter(javalang.tree.ClassDeclaration)
-        return [node._position.line for _, node in tree if [n for n in classes if n in node.name.lower()] != []]
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:27]: :+1: 
Glad we get rid of nested list comprehensions.

File: aibolit/patterns/er_class/er_class.py (hunk 6 above)
Comment (Vitaly-Protasov) [reply aibolit/patterns/er_class/er_class.py]: +

File: aibolit/patterns/er_class/er_class.py (hunk 3 above)
Comment (Vitaly-Protasov) [reply aibolit/patterns/er_class/er_class.py]: done

File: aibolit/patterns/er_class/er_class.py (hunk 2 above)
Comment (Vitaly-Protasov) [reply aibolit/patterns/er_class/er_class.py]: done

File: aibolit/patterns/er_class/er_class.py (hunk 1 above)
Comment (Vitaly-Protasov) [reply aibolit/patterns/er_class/er_class.py:22]: added

File: aibolit/patterns/er_class/er_class.py [hunk 8]
```diff
@@ -1,27 +1,57 @@
...
+    def value(self, filename: str) -> List[int]:
+        lines: List[int] = []
+        ast = AST.build_from_javalang(build_ast(filename))
+        for node in list(ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION)):
+            names = []
+            class_name = node.name.lower()
+            if any(forbiden_word in class_name for forbiden_word in self.forbiden_words_in_class_names):
+                names.append(class_name)
+            if names:
+                lines.append(node.line)
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:56]:
```suggestion
            class_name = node.name.lower()
            if any(forbidden_word in class_name for forbidden_word in self.forbiden_words_in_class_names):
                lines.append(node.line)
```

File: aibolit/patterns/er_class/er_class.py [hunk 9]
```diff
@@ -1,27 +1,57 @@
...
-                   'listener',
-                   'producer',
-                   'holder',
-                   'interceptor')
-        tree = build_ast(filename).filter(javalang.tree.ClassDeclaration)
-        return [node._position.line for _, node in tree if [n for n in classes if n in node.name.lower()] != []]
+    def value(self, filename: str) -> List[int]:
+        lines: List[int] = []
+        ast = AST.build_from_javalang(build_ast(filename))
+        for node in list(ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION)):
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:50]:
```suggestion
        for node in ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION):
```

File: aibolit/patterns/er_class/er_class.py [hunk 10]
```diff
@@ -1,27 +1,57 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from aibolit.ast_framework import ASTNodeType, AST
 from aibolit.utils.ast_builder import build_ast
+from typing import List
 
 
 class ErClass:
+    '''
+    Description: If a class name is one of the following (or ends with this word), it's the pattern
```
Comment (aravij) [inline aibolit/patterns/er_class/er_class.py:29]: We actually check if forbidden word is in class name, no matter in the beggining, end or in the middle.

**Comment by aravij** [comment]
@acheshkov  please merge

**Comment by acheshkov** [comment]
@rultor merge

**Comment by aravij** [comment]
Does it merged or not?

**Pull request merged**